### DISCUSSION TITLE
No suitable driver found for jdbc:sqlserver:// with SQL connection on initialize script

### DISCUSSION THREAD

**Discussion by spoekles** [eclipse-birt/birt#1412] [Q&A] [answered]
Hello, I am trying to connect to a database on the initialization phase of the report. I've managed to successfully make a connection when building the report through the Birt runtime (4.13.0.v202303021806). However when running the preview through the all-in-one editor (same version) the following error pops up

`org.eclipse.birt.report.engine.api.EngineException: Invalid javascript expression: Wrapped java.sql.SQLException: No suitable driver found for jdbc:sqlserver://...;databaseName=...;integratedSecurity=true;trustServerCertificate=true;authenticationScheme=JavaKerberos (/report/method[@name="initialize"]#120)` The dots for server/db are edited afterwards.

The script itself:

```
importPackage( Packages.java.sql )
importPackage( Packages.java.lang )

userDomain = java.lang.System.getenv("USERDNSDOMAIN");
userDomain = BirtStr.trim(userDomain);
userName = java.lang.System.getenv("USERNAME");
userName = BirtStr.trim(userName);
dataBase = reportContext.getParameterValue("DBselect");
fullUserName = userName + "@" + userDomain

var conn = DriverManager.getConnection("jdbc:sqlserver://...;databaseName=" + dataBase + ";integratedSecurity=true;trustServerCertificate=true;authenticationScheme=JavaKerberos","fullUserName","");
if (conn != null) 
{
    java.lang.System.out.println("Connected to " + dataBase);
}
```

I've tried enumerating the found drivers with

```
var drivers = DriverManager.getDrivers();

while (drivers.hasMoreElements()) {
    var driver = drivers.nextElement();
    java.lang.System.out.println(driver.getClass().getName());
}
```

The runtime through the console reports it can find `com.microsoft.sqlserver.jdbc.SQLServerDriver` but as far as I am aware I can't see the logs when previewing through the editor.

Any help would be greatly appreciated, if any more information is required please ask!

**Comment by speckyspooky** [comment]
You can try it with the class load of the MS SQL driver before your create a new connection.

Example of Java (like pattern example): 

```
import java.sql.*;
public class JDBCforName {
  
  public static void main(String[] args) {
    Connection connection = null;
      String url = "jdbc:sqlserver://localhost:1433/";
      String dbName = "db_name";
      String driverName = "com.microsoft.sqlserver.jdbc.SQLServerDriver";
      String userName = "root";
      String password = "root";
      try{
        Class.forName(driverName).newInstance();
        connection = DriverManager.getConnection(url+dbName, userName, password);
        System.out.println("java Class.forName(String drivername) Example");
      }
      catch (Exception e){
        e.printStackTrace();
      }      
  }
}

**Reply by spoekles** [reply]
Hi Specky, thanks for the reply, however the result doesn't change. The same fault code appears.

**Reply by speckyspooky** [reply, marked as answer]
I started a small research and the main problem is that the class path to the driver jar are not correct set to load the class.
So I tried a way to load the driver with an absolut path to the jar file. For the test I used BIRT 4.12 with Java 17.

### Result of full path load and SQL-statement (w/o special exception handling)
![2023-09-20 22_04_07-workspace - C__Users_gutma_Documents_Entwicklung_2_BIRT_report_dev_images_new_so](https://github.com/eclipse-birt/birt/assets/41593722/36acd7fe-de63-480d-8d29-fe55a35e406b)
### Demo-Report
[manual-driver-sql-demo.zip](https://github.com/eclipse-birt/birt/files/12677147/manual-driver-sql-demo.zip)

**Reply by spoekles** [reply]
Awesome, this works perfectly! Thank you very much 👍
